bank-vaults/vault-secrets-reloader
# POC

A first prototype for vault-secrets-reloader for deployments, findings and issues are addressed in issue #1 .

### Steps for trying out locally

1. Prepare a kubernetes cluster (recommended with Docker desktop or kind)
2. Start the operator either with `make up` from the operator repo locally, or via Helm:
```
helm upgrade --install vault-operator ~/Github/bank-vaults/vault-operator/deploy/charts/vault-operator \
    --set image.tag=latest \
    --set image.bankVaultsTag=latest \
    --set image.pullPolicy=IfNotPresent \
    --wait
```
3. Create the namespace `vswh` (`kubectl create namespace vswh`)
4. Create a Vault instance with `kubectl apply -f test-cr.yaml`
5. Install the webhook:
```
helm upgrade --install vault-secrets-webhook charts/vault-secrets-webhook \
    --set replicaCount=1 \
    --set image.tag=local \
    --set image.pullPolicy=IfNotPresent \
    --set configMapMutation=true \
    --set deploymentSecretSync=true \
    --set configmapFailurePolicy=Fail \
    --set podsFailurePolicy=Fail \
    --set secretsFailurePolicy=Fail \
    --set vaultEnv.tag=latest \
    --namespace vswh
```
6. Apply the test deployment: `kubectl apply -f test/deploy/test-deployment-reload.yaml`
7. Build the reloader image (`make docker-build`) and start it with `make deploy`
8. Check if the secret `vault-secrets-reloader-hashes` in namespace `vswh` got been created and filled with data
9. Start port-forwarding Vault: `kubectl port-forward vault-0 8200:8200` and in another terminal instance, export some important envs that the CLI tool needs:
```
export VAULT_TOKEN=$(kubectl get secrets vault-unseal-keys -o jsonpath={.data.vault-root} | base64 --decode)

kubectl get secret vault-tls -o jsonpath="{.data.ca\.crt}" | base64 --decode > $PWD/vault-ca.crt
export VAULT_CACERT=$PWD/vault-ca.crt

export VAULT_ADDR=https://127.0.0.1:8200
```
10. Now change a secret in Vault in the same terminal window: `vault kv patch secret/mysql MYSQL_PASSWORD=updated3xtr3ms3cr3t`
11. The test deployment should be restarted soon, with the new secret is present in its logs
12. You can clean up after the reloader with `make undeploy`
